cpp/src/fil/treelite_import.cu

```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
   return 0;  // Treelite format assumes that the root is 0
 }
 
-template <typename T, typename L>
-inline int max_depth(const tl::Tree<T, L>& tree)
+struct empty {
+};
+
+/** walk a Treelite tree, visiting each inner node with inner_func and each leaf node with
+  leaf_func. See walk_tree::invitation::state documentation for how RootToLeafPathState is retained
+during traversal. Any per-tree state during traversal should be captured by the lambdas themselves.
+  inner_func(int node_id, RootToLeafPathState state) should return a pair of new states, one for
+each child node. leaf_func(int, RootToLeafPathState) returns nothing.
+**/
+template <typename RootToLeafPathState,
+          typename Threshold,
+          typename Leaf,
+          typename InnerNodeVisitFunc,
+          typename LeafNodeVisitFunc = decltype([](int, RootToLeafPathState) {})>
+inline void walk_tree(
+  const tl::Tree<Threshold, Leaf>& tree,
+  InnerNodeVisitFunc inner_func,
+  LeafNodeVisitFunc leaf_func = [](int, RootToLeafPathState) {})
 {
-  // trees of this depth aren't used, so it most likely means bad input data,
-  // e.g. cycles in the forest
-  const int DEPTH_LIMIT = 500;
-  int root_index        = tree_root(tree);
-  typedef std::pair<int, int> pair_t;
-  std::stack<pair_t> stack;
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
+  /// needed to visit a node
+  struct invitation {
```

Comment:
Not sure `invitation` is a good name, unless it's a standard term for such functions (`visit()` is definitely a standard term). I think `element` would be better.